rpc.go

```diff
@@ -39,49 +39,76 @@ import (
 
 type (
 	rpcClient struct {
-		ec                 entityConnector
-		client             *amqp.Client
-		clientMu           sync.RWMutex
+		ec     entityConnector
+		client *amqp.Client
+
+		clientMu  sync.RWMutex
+		linkCache map[string]*rpc.Link // stores 'address' => rpc.Link.
+
 		sessionID          *string
 		isSessionFilterSet bool
 		cancelAuthRefresh  func() <-chan struct{}
+
+		// replaceable for testing
+
+		// alias of 'rpc.NewLink'
+		newRPCLink func(conn *amqp.Client, address string, opts ...rpc.LinkOption) (*rpc.Link, error)
+
+		// alias of function 'newAMQPClient'
+		newAMQPClient func(ctx context.Context, ec entityConnector) (*amqp.Client, func() <-chan struct{}, error)
+
+		// alias of 'amqp.Client.Close()'
+		closeAMQPClient func() error
 	}
 
 	rpcClientOption func(*rpcClient) error
 )
 
 func newRPCClient(ctx context.Context, ec entityConnector, opts ...rpcClientOption) (*rpcClient, error) {
 	r := &rpcClient{
-		ec: ec,
+		ec:            ec,
+		linkCache:     map[string]*rpc.Link{},
+		newRPCLink:    rpc.NewLink,
+		newAMQPClient: newAMQPClient,
 	}
 
+	r.closeAMQPClient = func() error { return r.client.Close() }
+
 	for _, opt := range opts {
 		if err := opt(r); err != nil {
 			tab.For(ctx).Error(err)
 			return nil, err
 		}
 	}
-	if err := r.newClient(ctx); err != nil {
+
+	var err error
+	r.client, r.cancelAuthRefresh, err = r.newAMQPClient(ctx, r.ec)
+
+	if err != nil {
 		tab.For(ctx).Error(err)
 		return nil, err
 	}
 	return r, nil
 }
 
-// newClient will replace the existing client and start auth auto-refresh.
-// any pre-existing client MUST be closed before calling this method.
-// NOTE: this does *not* take the write lock, callers must hold it as required!
-func (r *rpcClient) newClient(ctx context.Context) error {
-	var err error
-	r.client, err = r.ec.Namespace().newClient(ctx)
+// newAMQPClient creates a client and starts auth auto-refresh.
+// NOTE: this function can "partially" succeed. The returned amqp.Client should always be closed,
+// even when an error is returned.
+func newAMQPClient(ctx context.Context, ec entityConnector) (*amqp.Client, func() <-chan struct{}, error) {
+	client, err := ec.Namespace().newClient(ctx)
 	if err != nil {
-		return err
+		return nil, nil, err
 	}
-	r.cancelAuthRefresh, err = r.ec.Namespace().negotiateClaim(ctx, r.client, r.ec.ManagementPath())
+
+	cancelAuthRefresh, err := ec.Namespace().negotiateClaim(ctx, client, ec.ManagementPath())
+
 	if err != nil {
-		return err
+		// it's important to return the `client` here so the calling code
```

Comment:
I think we should just close the client here.  The idiom is, whenever the returned error != `nil` you shouldn't touch the other things being returned.

Reply:
Good point, will change.

Reply:
Changed. It's annoying how I can't test it currently (`Namespace()` is publicly exposed through the Subscription and Queue structs because of embedding).